### PR TITLE
Fix focusscope restore logic

### PR DESCRIPTION
**Pull request by snowystinger**
Closes https://github.com/adobe/react-spectrum/issues/4514

After reviews, we determined the root cause was even further down. Comments in the code should now provide the explanation. There is no test because it would require mocking rafs such that we control the order in which they are run because this is a race condition and it just does not exist in the fake timers implementation we currently have with out doing the difficult extra work listed. Picker tests seem to cover cases close to this and no other tests were affected.

~See comment https://github.com/adobe/react-spectrum/issues/4514#issuecomment-1734883106 for why i think this is the correct fix~

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

### CODE REVIEW

File: packages/@react-spectrum/list/test/ListView.test.js [hunk 1]
```diff
@@ -604,6 +610,68 @@ describe('ListView', function () {
       expect(document.activeElement).toBe(rows[rows.length - 2]);
       expect(rows[rows.length - 1]).toHaveAttribute('aria-disabled', 'true');
     });
+
+    // TODO: fix this test, it appears that the table item animation out isn't running
```
Comment (iamwillpowell) [inline packages/@react-spectrum/list/test/ListView.test.js:614]: Is this tracked with a separate issue?

**Review by iamwillpowell — APPROVED** [review]

File: packages/@react-spectrum/list/test/ListView.test.js (hunk 1 above)
Comment (snowystinger) [reply packages/@react-spectrum/list/test/ListView.test.js:614]: hoping someone has an idea while reviewing honestly

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 2]
```diff
@@ -131,43 +131,42 @@ export function FocusScope(props: FocusScopeProps) {
...
       }
+    }
+  }, [scopeRef]);
 
-      return () => {
-        // Scope may have been re-parented.
-        let parentScope = focusScopeTree.getTreeNode(scopeRef)?.parent?.scopeRef ?? null;
+  // This layout effect cleanup is so that the tree node is removed synchronously with react before the RAF
+  // in useRestoreFocus cleanup runs.
+  useLayoutEffect(() => {
```
Comment (LFDanLu) [inline packages/@react-aria/focus/src/FocusScope.tsx:157]: So to make sure I understand the previous issue:
We previously handled the cleanup in a useEffect cleanup and that could potentially run after the RAF triggered by the useRestoreFocus useLayoutEffect cleanup?

Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:157]: Yep, it was a race for which got scheduled first, the cleanup raf or the useeffect raf.

**Review by LFDanLu — APPROVED** [review]
🤞

**Review by devongovett — APPROVED** [review]

**Pull request merged**